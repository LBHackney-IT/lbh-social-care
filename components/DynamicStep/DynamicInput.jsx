import PropTypes from 'prop-types';

import * as Inputs from 'components/Form';

const DynamicInput = ({
  component,
  register,
  control,
  id,
  name,
  errors,
  multiStepIndex,
  ...otherProps
}) => {
  const inputName =
    typeof multiStepIndex === 'number'
      ? `${id}[${multiStepIndex}].${name}`
      : name;
  const Component = Inputs[component];
  const sharedProps = {
    name: inputName,
    error: errors[inputName],
    required: otherProps?.rules?.required,
    ...otherProps,
  };
  switch (component) {
    case 'AddressLookup':
    case 'DateInput':
      return <Component control={control} {...sharedProps} />;
    default:
      return (
        <Component register={register(otherProps.rules)} {...sharedProps} />
      );
  }
};

DynamicInput.propTypes = {
  component: PropTypes.string.isRequired,
  register: PropTypes.func.isRequired,
  control: PropTypes.object.isRequired,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  errors: PropTypes.object.isRequired,
  rules: PropTypes.shape({
    required: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  }),
  multiStepIndex: PropTypes.number,
};

export default DynamicInput;
