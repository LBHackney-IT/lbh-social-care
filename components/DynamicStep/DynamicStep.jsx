import PropTypes from 'prop-types';
import { useForm } from 'react-hook-form';

import { Button } from 'components/Form';
import DynamicInput from 'components/DynamicInput/DynamicInput';

const DynamicStep = ({ components, formData, onStepSubmit, onSaveAndExit }) => {
  const { handleSubmit, register, control, errors } = useForm({
    defaultValues: formData,
  });
  return (
    <>
      <form>
        <div className="govuk-form-group">
          {components?.map((componentProps) =>
            componentProps.name ? (
              <DynamicInput
                key={componentProps.name}
                register={register}
                control={control}
                errors={errors}
                {...componentProps}
              />
            ) : (
              componentProps
            )
          )}
        </div>
        <Button
          className="govuk-button"
          label="Next"
          type="submit"
          onClick={handleSubmit(onStepSubmit)}
        />
        <Button
          className="govuk-button"
          label="Save and Exit"
          type="submit"
          onClick={handleSubmit(onSaveAndExit)}
        />
      </form>
    </>
  );
};

DynamicStep.propTypes = {
  components: PropTypes.array,
  onStepSubmit: PropTypes.func.isRequired,
  formData: PropTypes.object.isRequired,
  isLoading: PropTypes.bool,
  hasError: PropTypes.bool,
};

export default DynamicStep;
