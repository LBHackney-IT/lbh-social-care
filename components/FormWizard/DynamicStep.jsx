import { isValidElement, useCallback } from 'react';
import PropTypes from 'prop-types';
import { useForm } from 'react-hook-form';
import cx from 'classnames';

import Button from 'components/Button/Button';
import DynamicInput from 'components/FormWizard/DynamicInput';
import DynamicInputMulti from 'components/FormWizard/DynamicInputMulti';

import styles from './DynamicStep.module.scss';

const DynamicStep = ({
  isMulti,
  stepId,
  components,
  formData,
  onStepSubmit,
  onSaveAndExit,
}) => {
  const { handleSubmit, register, control, errors, setValue, watch } = useForm({
    defaultValues: formData,
  });
  const stepValues = watch();
  const currentData = {
    ...formData,
    ...stepValues,
  };
  const multiStepPrefix =
    isMulti && `${stepId[0]}[${parseInt(stepId[1]) - 1 || 0}]`;
  const sanitiseData = useCallback(
    (data) => ({
      ...components
        .filter(({ isMulti }) => isMulti)
        .reduce((acc, { name }) => ({ ...acc, [name]: undefined }), {}),
      ...data,
    }),
    [components]
  );
  if (!register) {
    return null;
  }
  return (
    <>
      <form
        role="form"
        onSubmit={handleSubmit((data) => onStepSubmit(sanitiseData(data)))}
      >
        <div className="govuk-form-group">
          {components?.map(
            ({
              conditionalRender,
              showConditionalGuides,
              name,
              isMulti: isComponentMulti,
              ...componentProps
            }) => {
              if (isValidElement(componentProps)) {
                return componentProps;
              }
              if (conditionalRender && !conditionalRender(currentData)) {
                return null;
              }
              const inputName = multiStepPrefix
                ? `${multiStepPrefix}.${name}`
                : name;
              const sharedProps = {
                key: inputName,
                name: inputName,
                register: register,
                control: control,
                errors: errors,
                currentData: currentData,
                ...componentProps,
              };
              return (
                <div
                  key={inputName}
                  className={cx('govuk-form-group', {
                    [styles.withConditionalGuides]:
                      conditionalRender && showConditionalGuides,
                  })}
                >
                  {isComponentMulti ? (
                    <DynamicInputMulti
                      {...sharedProps}
                      initialInputData={formData[name]}
                      onDelete={(updatedValue) =>
                        setValue(inputName, updatedValue)
                      }
                    />
                  ) : (
                    <DynamicInput {...sharedProps} />
                  )}
                </div>
              );
            }
          )}
        </div>
        {isMulti && (
          <Button
            isSecondary
            label="Add another"
            type="button"
            onClick={() =>
              handleSubmit((data) => onStepSubmit(sanitiseData(data), true))()
            }
          />
        )}
        <div className={styles.actions}>
          <Button
            className="lbh-button govuk-button"
            isSecondary
            wideButton
            label="Save and finish later"
            type="button"
            onClick={() => onSaveAndExit(stepValues)}
          />
          <a
            href="/"
            className="lbh-link govuk-link"
            style={{
              marginLeft: '1.5rem',
            }}
          >
            Cancel
          </a>

          <Button label="Continue" type="submit" />
        </div>
      </form>
    </>
  );
};

DynamicStep.propTypes = {
  stepId: PropTypes.array,
  components: PropTypes.array,
  onStepSubmit: PropTypes.func.isRequired,
  onSaveAndExit: PropTypes.func.isRequired,
  formData: PropTypes.object.isRequired,
  isMulti: PropTypes.bool,
};

export default DynamicStep;
