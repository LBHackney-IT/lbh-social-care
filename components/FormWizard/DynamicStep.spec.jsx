import { render, fireEvent, waitFor } from '@testing-library/react';

import DynamicStep from './DynamicStep';

describe('DynamicStep component', () => {
  const props = {
    stepId: ['Foo'],
    components: [
      {
        component: 'TextInput',
        name: 'bar_input',
        width: 30,
        label: 'Foo',
      },
    ],
    formData: {
      id: 44000,
      bar_input: 'foo',
    },
    onStepSubmit: jest.fn(),
    onSaveAndExit: jest.fn(),
  };
  it('should render properly', () => {
    const { asFragment } = render(<DynamicStep {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should handle next button', async () => {
    const { getByRole } = render(<DynamicStep {...props} />);
    const input = getByRole('textbox', { name: 'Foo' });
    const nextButton = getByRole('button', { name: 'Continue' });

    expect(input).toBeInTheDocument();
    expect(nextButton).toBeInTheDocument();

    fireEvent.input(input, { target: { value: 'bar foo' } });
    fireEvent.submit(nextButton);

    await waitFor(() => expect(props.onStepSubmit).toHaveBeenCalled());
    expect(props.onStepSubmit).toHaveBeenCalledWith({
      bar_input: 'bar foo',
    });
  });

  it('should fire save and exit button', async () => {
    const { getByRole } = render(<DynamicStep {...props} />);
    const saveButton = getByRole('button', { name: 'Save and finish later' });

    expect(saveButton).toBeInTheDocument();

    fireEvent.click(saveButton);

    await waitFor(() => expect(props.onSaveAndExit).toHaveBeenCalledTimes(1));
  });
});
