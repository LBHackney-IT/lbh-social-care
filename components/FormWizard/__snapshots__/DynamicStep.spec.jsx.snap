// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicStep component should render properly 1`] = `
<DocumentFragment>
  <form
    role="form"
  >
    <div
      class="govuk-form-group"
    >
      <div
        class="govuk-form-group"
      >
        <div
          class="govuk-form-group lbh-form-group"
        >
          <label
            class="govuk-label lbh-label"
            for="bar_input"
          >
            Foo 
          </label>
          <input
            class="govuk-input lbh-input govuk-input--width-30"
            data-testid="bar_input"
            id="bar_input"
            name="bar_input"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      class="actions"
    >
      <button
        class="lbh-button govuk-button lbh-button--secondary lbh-button govuk-button"
        data-module="govuk-button"
        style="min-width: 200px;"
        type="button"
      >
        Save and finish later
      </button>
      <a
        class="lbh-link govuk-link"
        href="/"
        style="margin-left: 1.5rem;"
      >
        Cancel
      </a>
      <button
        class="lbh-button govuk-button"
        data-module="govuk-button"
        type="submit"
      >
        Continue
      </button>
    </div>
  </form>
</DocumentFragment>
`;
